### PR TITLE
examples: refine tensor_sum_elements(tensor dump) in examples/benchmark/benchmark-matmult.cpp

### PR DESCRIPTION
**Pull request by zhouwg**
Self Reported Review Complexity

- [x] Review Complexity : Low
- [ ] Review Complexity : Medium
- [ ] Review Complexity : High
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)

BTW, can the original author or core maintainer add a 
```
void ggml_tensor_print(const struct ggml_tensor * tensor, const char * name) 

```
utility function (with all supported data type and dump the accurate value in the tensor as m ( < 8) x n ( < 8) format) in ggml.h&ggml.c, it will might be very useful/helpful for programmers/developers. thanks so much.

### CODE REVIEW

File: examples/benchmark/benchmark-matmult.cpp [hunk 1]
```diff
@@ -31,15 +32,58 @@ static void ggml_graph_compute_helper(std::vector<uint8_t> & buf, ggml_cgraph *
...
+        }
+    }
+
+    if (tensor->type == GGML_TYPE_Q8_0) {
+        int blocks = 0;
+        block_q8_0 * quant_datas = (block_q8_0 *)tensor->data;
+        for (int j = 0; j < tensor->ne[1]; j++) {
+            blocks = tensor->ne[0] / QK8_0;
+            for (int i = 0; i < blocks; i++) {
+                floatvalue = GGML_FP16_TO_FP32(quant_datas[j * blocks + i].d);
```
Comment (ngxson) [inline examples/benchmark/benchmark-matmult.cpp:79]: This is not correct for converting q8_0 to fp32. To convert any q type to fp32 (except for IQ), you can do:

```ggml_type_traits_t qtype = ggml_internal_get_type_traits(tensor->type);
qtype.to_float(data, f32_output, nelements);
```

Comment (zhouwg) [reply examples/benchmark/benchmark-matmult.cpp:79]: thanks your correction.I'll study and modify it accordingly.

Comment (ngxson) [reply examples/benchmark/benchmark-matmult.cpp:94]: I'm not sure either, but it seems to me that the `dequantize_row_q8_0` calculates `const int nb = k / qk;` with k is the total number of elements, while in your implementation k = `tensor->ne[0]`

File: examples/benchmark/benchmark-matmult.cpp [hunk 2]
```diff
@@ -31,15 +32,74 @@ static void ggml_graph_compute_helper(std::vector<uint8_t> & buf, ggml_cgraph *
...
+            printf("malloc failed\n");
+            return 0.0;
+        }
+        qtype.to_float(quant_datas, float32, tensor->ne[0] * tensor->ne[1]);
+        for (int j = 0; j < tensor->ne[1]; j++) {
+            for (int k = 0; k < tensor->ne[0]; k++) {
+                sum += float32[j * tensor->ne[0] + k];
             }
         }
+        free(float32);
```
Comment (ngxson) [inline examples/benchmark/benchmark-matmult.cpp:88]: You can get rid of `malloc` and all instances of `tensor->ne`:

```cpp
std::vector<float> f32out(ggml_nelements(tensor));
ggml_type_traits_t qtype = ggml_internal_get_type_traits(tensor->type);
qtype.to_float((void *) tensor->data, f32out.data(), f32out.size());
for (float e : f32out) sum += e;
```

It also works with other types of want, so maybe we can replace this whole function with this version.

Please note that I didn't test it, so up to you to test / adapt.

Comment (zhouwg) [reply examples/benchmark/benchmark-matmult.cpp:88]: excellent!

File: examples/benchmark/benchmark-matmult.cpp (hunk 1 above)
Comment (zhouwg) [reply examples/benchmark/benchmark-matmult.cpp:94]: k = tensor->ne[0] is used to calculate block counts.

File: examples/benchmark/benchmark-matmult.cpp [hunk 3]
```diff
@@ -32,14 +33,45 @@ static void ggml_graph_compute_helper(std::vector<uint8_t> & buf, ggml_cgraph *
...
     if (tensor->type == GGML_TYPE_F32) {
         for (int j = 0; j < tensor->ne[1]; j++) {
             for (int k = 0; k < tensor->ne[0]; k++) {
-                sum += ((float *) tensor->data)[j*tensor->ne[0] + k];
+                sum += ((float *) tensor->data)[j * tensor->ne[0] + k];
             }
         }
     }
+
+    if (ggml_is_quantized(tensor->type)) {
```
Comment (ngxson) [inline examples/benchmark/benchmark-matmult.cpp:66]: `GGML_TYPE_I8` is also a quantized type. All `if` conditions in this function does not have `else`, so the sum will be calculated twice for some types.

**Review by ngxson — CHANGES_REQUESTED** [review]

File: examples/benchmark/benchmark-matmult.cpp (hunk 3 above)
Comment (zhouwg) [reply examples/benchmark/benchmark-matmult.cpp]: thanks!

**Review by ngxson — APPROVED** [review]
LGTM. Let's merge when the CI pass.

**Review by slaren — CHANGES_REQUESTED** [review]
What is the goal of these changes when this function is only called with F32 tensors?

**Comment by zhouwg** [comment]
> What is the goal of these changes when this function is only called with F32 tensors?

programmer/developer might be change data type in this example manually.

**Comment by slaren** [comment]
There is code for printing tensors in the `eval-callback` example. We could consider moving and possibly extending the functions used to do this in `eval-callback` to `common.cpp` so that it is easier to use them during debugging.

**Comment by zhouwg** [comment]
you can make/made any decision as your consideration because you are one of the owners/core maintainers  of this project.

**Comment by zhouwg** [comment]
you can make/made any decision as your consideration because you are one of the owners/core maintainers  of this project.

> LGTM. Let's merge when the CI pass.

thanks for your help and thanks so much and have a good weekend.

**Comment by ngxson** [comment]
Good point @slaren , I haven’t looked deeper into this file.

@zhouwg I’m a bit doubt about the purpose of this function. Just to be sure: do you need this func to print the tensor? It seems to me that the current func only calculate the sum, but not for printing it out.

**Comment by zhouwg** [comment]
> Good point @slaren , I haven’t looked deeper into this file.
> 
> @zhouwg I’m a bit doubt about the purpose of this function. Just to be sure: do you need this func to print the tensor? It seems to me that the current func only calculate the sum, but not for printing it out.

your are right.this function following the existing tensor_sum_elements but the calculation process can be used for tensor dump,

**Pull request closed without merging**